drivers/wifi/esp8266/wifi_esp8266.c

```diff
@@ -0,0 +1,1452 @@
+			  slen);
+	result.ssid_length = slen;
+
+	/* rssi */
+	slen = delimiters[3] - delimiters[2];
+	net_buf_linearize(temp, 32, *buf, delimiters[2], slen);
+	temp[slen] = '\0';
+	result.rssi = strtol(temp, NULL, 10);
+
+	/* channel */
+	slen = delimiters[5] - delimiters[4];
+	net_buf_linearize(temp, 32, *buf, delimiters[4], slen);
+	temp[slen] = '\0';
+	result.channel = strtol(temp, NULL, 10);
+
+	/* issue callback to report scan results */
+	if (foo_data.wifi_scan_cb) {
+		foo_data.wifi_scan_cb(foo_data.iface, 0, &result);
+	}
+}
+
+static const char nm_label[] = "netmask";
+static const char gw_label[] = "gateway";
+static const char ip_label[] = "ip";
+
+static void on_cmd_ip_addr_resp(struct net_buf **buf, u16_t len)
+{
+	char ip_addr[] = "xxx.xxx.xxx.xxx";
+	int d[3];
+	int slen;
+
+	d[0] = net_buf_find_next_delimiter(*buf, ':', 0);
+	d[1] = net_buf_find_next_delimiter(*buf, '\"', d[0] + 1);
+	d[2] = net_buf_find_next_delimiter(*buf, '\"', d[1] + 1);
+
+	slen = d[2] - d[1] - 1;
+
+	net_buf_linearize(ip_addr, 16, *buf, d[0] + 2, slen);
+	ip_addr[slen] = '\0';
+
+	if (net_buf_ncmp(*buf, nm_label, strlen(nm_label)) == 0) {
+		net_addr_pton(AF_INET, ip_addr, &foo_data.netmask);
+	} else if (net_buf_ncmp(*buf, ip_label, strlen(ip_label)) == 0) {
+		net_addr_pton(AF_INET, ip_addr, &foo_data.ip);
+	} else if (net_buf_ncmp(*buf, gw_label, strlen(gw_label)) == 0) {
+		net_addr_pton(AF_INET, ip_addr, &foo_data.gw);
+	} else {
+		return;
+	}
+}
+
+static void on_cmd_mac_addr_resp(struct net_buf **buf, u16_t len)
+{
+	int i;
+	char octet[] = "xx";
+	for (i = 0; i < 6; i++) {
+		net_buf_pull_u8(*buf);
+		octet[0]= net_buf_pull_u8(*buf);
+		octet[1]= net_buf_pull_u8(*buf);
+		foo_data.mac[i] = strtoul(octet, NULL, 16);
+	}
+}
+
+static void on_cmd_sendok(struct net_buf **buf, u16_t len)
+{
+	foo_data.last_error = 0;
+
+	/* only signal semaphore if we are not waiting on a data send command */
+	k_sem_give(&foo_data.response_sem);
+}
+
+/* Handler: OK */
+static void on_cmd_sockok(struct net_buf **buf, u16_t len)
+{
+	foo_data.last_error = 0;
+
+	if (!foo_data.sending_data) {
+		k_sem_give(&foo_data.response_sem);
+	}
+}
+
+/* Handler: ERROR */
+static void on_cmd_sockerror(struct net_buf **buf, u16_t len)
+{
+	foo_data.last_error = -EIO;
+	k_sem_give(&foo_data.response_sem);
+}
+
+static void on_cmd_wifi_connected_resp(struct net_buf **buf, u16_t len)
+{
+	wifi_mgmt_raise_connect_result_event(foo_data.iface, 0);
+}
+
+static void on_cmd_wifi_disconnected_resp(struct net_buf **buf, u16_t len)
+{
+	wifi_mgmt_raise_disconnect_result_event(foo_data.iface, 0);
+}
+
+static int esp8266_get(sa_family_t family,
+			enum net_sock_type type,
+			enum net_ip_protocol ip_proto,
+			struct net_context **context)
+{
+	int i;
+	struct net_context *c = *context;
+
+	if (family != AF_INET) {
+		return -1;
+	}
+
+	k_sem_take(&sock_sem, K_FOREVER);
+	for (i = 0; i < ESP8266_MAX_CONNECTIONS && foo_data.sock_map & BIT(i); i++){}
```

Comment:
Instead of this, perhaps you could use the `u32_count_leading_zeros()` to calculate the bits.
Also there is the `ATOMIC_DEFINE()` macro that is used for example in ethernet.h, to store a bit-map of the VLAN interfaces. Sounds pretty similar use case what you have here.